### PR TITLE
💚 log failed e2e resources and related events

### PR DESCRIPTION
**Pull request by mboersma**
**What this PR does / why we need it**:

Logs failed deployments (or jobs or services) and related events in the load balancer e2e test, hopefully to shine a light on some failures we've seen in CI.

**Which issue(s) this PR fixes**:

Refs #807
Refs #814

**Special notes for your reviewer**:

This copies some code from CAPI e2e helpers temporarily, rather than vendor in CAPI at an arbitrary commit.

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
 🌱 [e2e] log failed deployments and increase HTTP retries in LB test
```

### CODE REVIEW

File: test/e2e/azure_lb.go
Comment (mboersma) [inline test/e2e/azure_lb.go:275]: Maybe this should call `DescribeFailedService` instead, after #860 merges.

**Comment by CecileRobertMichon** [comment]
/lgtm
/approve
/hold

Is the same change available in CAPI v0.3.9? If so that should be out in the next couple of days

**Comment by nader-ziada** [comment]
> Is the same change available in CAPI v0.3.9? If so that should be out in the next couple of days

CAPI only has the Deployments helper, but this includes also the Service and Job helpers

**Comment by mboersma** [comment]
> CAPI only has the Deployments helper, but this includes also the Service and Job helpers

And this change to the Deployments helper isn't in CAPI (yet). I can PR that change against CAPI first if we'd prefer, because this does diverge a fair amount.

**Comment by nader-ziada** [comment]
> > CAPI only has the Deployments helper, but this includes also the Service and Job helpers
> 
> And this change to the Deployments helper isn't in CAPI (yet). I can PR that change against CAPI first if we'd prefer, because this does diverge a fair amount.

I'm okay with diverging and not waiting now that the CAPI release is out, but lets see what @CecileRobertMichon thinks

**Comment by CecileRobertMichon** [comment]
Now that #918 is open, let's just clean this up to include anything that's not in v0.3.9

**Comment by mboersma** [comment]
/retest

It has failed to provision all 3 control-plane nodes twice now. Boo!

**Comment by mboersma** [comment]
@CecileRobertMichon I think this is ready. It changes some CAPZ test helper functions and clones the CAPI `WaitForDeploymentsAvailable` function locally in order to make some changes to it. I'll PR those back upstream if we think this is good.

I've seen a couple e2e runs lately where this might have pointed out why an LB test failed.

File: test/e2e/azure_lb.go
Comment (CecileRobertMichon) [inline test/e2e/azure_lb.go:133]: why this change?

File: test/e2e/azure_lb.go
Comment (CecileRobertMichon) [inline test/e2e/azure_lb.go:208]: same question

File: test/e2e/azure_lb.go
Comment (mboersma) [reply test/e2e/azure_lb.go:133]: I think that's a holdover from when I was trying to fix the LB e2e spec failure, and since we don't use the HTTPS port here I thought to simplify things.

Comment (CecileRobertMichon) [reply test/e2e/azure_lb.go:133]: let's either remove it or make it a separate commit (in this PR) to make it clear there was a behavior change beyond logging

Comment (mboersma) [reply test/e2e/azure_lb.go]: I removed that change (restored the https:// port).

**Review by CecileRobertMichon — APPROVED** [review]
lgtm

let's bring these changes back into the framework as well.

@mboersma can you please fix the conflict then we'll merge it and I'll rebase #646 on top of both commits

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm

**Pull request merged**
